Add EIP-6105: Marketplace extension for EIP-721 (#6105)
ethereum/EIPs: EIPS/eip-6105.md

---
eip: 6105
title: Marketplace Extension for EIP-721
description: Adds a basic marketplace functionality to EIP-721.
author: Silvere Heraudeau (@lambdalf-dev), Martin McConnell (@offgridgecko), 5660-eth (@5660-eth), Abu <team10kuni@gmail.com>, Wizard Wang
discussions-to: https://ethereum-magicians.org/t/idea-a-marketplace-extension-to-erc721-standard/11975
status: Draft
type: Standards Track
category: ERC
created: 2022-12-02
requires: 165, 721
---

## Abstract

Add a basic marketplace functionality to [EIP-721](./eip-721.md)
to enable non-fungible token trading without relying on an intermediary trading platform.

## Motivation

Most current NFT trading relies on an NFT trading platform acting as an intermediary, which has the following problems:

> 1. Security issues. For example, security issues caused by authorization through `setApprovalForAll`. The permission obtained by the NFT trading platform has carried out unnecessary risk exposure. Once there is a problem with the trading platform contract, it will bring huge losses to the entire industry.If the user has authorized the operation right of his/her NFTs to the NFT trading platform, the latest phishing website fraud method is to trick the user into signing, and let the victim place an order at a very low price on the NFTs trading platform, and designate the recipient. The scammer then accepts the deal, thereby reaping illicit benefits. Ordinary users are hard-pressed to guard against this.
> 2. High trading costs. On the one hand, with the increase in the number of trading platforms, the liquidity of NFTs is scattered. If a user needs to make a deal as quickly as possible, he needs to authorize and place orders on multiple platforms, which further increases the risk exposure, and each authorization needs to spend gas. Taking BAYC as an example, the total number of BAYC is 10,000, and the number of current holders exceeds 6,000. The average number of BAYC held by each holder is less than 2. Although `setApprovalForAll` saves the subsequent pending order gas expenditure of a single trading platform, due to the need to authorize to Multi-platform, in essence, leads to an increase in gas expenditures for users. On the other hand, more importantly, the trading service fee charged by the trading platform must also be listed as the cost of user trading, which is much higher than the required gas spending for authorization.
> 3. The aggregator provides a solution to aggregate liquidity, but whether the aggregator aggregates the liquidity of a certain trading platform, its decision is centralized. And because the order information of the trading platform is off-chain, the efficiency of the aggregator to obtain order data is affected by the frequency of the trading platform API, and sometimes the NFT trading platform will suspend the distribution of APIs and limit the frequency of APIs.
> 4. Whether the NFT project party’s copyright tax income is obtained depends on the centralized decision-making of the NFT trading platform. Some trading platforms ignore the interests of the project party and implement 0 copyright tax, which infringes the interests of the project party. 
> 5. Not resistant to censorship. Some platforms have delisted many NFTs, and the formulation and implementation of delisting rules are centralized and not transparent enough. Previously, some NFT trading platforms have also failed and wrongly delisted some NFTs, which caused market panic.

Reviewer: There are a few grammatical issues with this. This is my best attempt at rephrasing this:
```suggestion
1. Security concerns arise from authorization via the `setApprovalForAll` function. The permissions granted to NFT trading platforms expose unnecessary risks. Should a problem occur with the trading platform contract, it would result in significant losses to the industry as a whole. Additionally, if a user has authorized the trading platform to handle their NFTs, it allows a phishing scam to trick the user into signing a message that allows the scammer to place an order at a low price on the NFT trading platform and designate themselves as the recipient. This can be difficult for ordinary users to guard against.
2. High trading costs are a significant issue. On one hand, as the number of trading platforms increases, the liquidity of NFTs becomes dispersed. If a user needs to make a deal quickly, they must authorize and place orders on multiple platforms, which increases the risk exposure and requires additional gas expenditures for each authorization. For example, taking BAYC as an example, with a total supply of 10,000 and over 6,000 current holders, the average number of BAYC held by each holder is less than 2. While `setApprovalForAll` saves on gas expenditure for pending orders on a single platform, authorizing multiple platforms results in an overall increase in gas expenditures for users. On the other hand, trading service fees charged by trading platforms must also be considered as a cost of trading, which are often much higher than the required gas expenditures for authorization.
3. Aggregators provide a solution by aggregating liquidity, but the decision-making process is centralized. Furthermore, as order information on trading platforms is off-chain, the aggregator's efficiency in obtaining data is affected by the frequency of the trading platform's API and, at times, trading platforms may suspend the distribution of APIs and limit their frequency.
4. The project parties' copyright tax income is dependent on centralized decision-making by NFT trading platforms. Some trading platforms disregard the interests of project parties and implement zero copyright tax, which is a violation of their interests.
5. NFT trading platforms are not resistant to censorship. Some platforms have delisted a number of NFTs and the formulation and implementation of delisting rules are centralized and not transparent enough. In the past, some NFT trading platforms have failed and wrongly delisted certain NFTs, leading to market panic.
```

Reviewer 2: Nice job.Thanks.

---
eip: 6105

> title: Marketplace extension for EIP-721

Reviewer: ```suggestion
title: Marketplace Extension for EIP-721
```
Should be title case

> description: Add a basic marketplace functionality to the EIP-721.

Reviewer: ```suggestion
description: Adds a basic marketplace functionality to EIP-721.
```

author: Silvere Heraudeau (@lambdalf-dev), Martin McConnell (@offgridgecko), 5660-eth (@5660-eth), Abu <team10kuni@gmail.com>, Wizard Wang
discussions-to: https://ethereum-magicians.org/t/idea-a-marketplace-extension-to-erc721-standard/11975
status: Draft
type: Standards Track
category: ERC
created: 2022-12-02
requires: 165, 721
---

> 
> "Not your marketplace, not your royalties"

Reviewer: ```suggestion
```
@SamWilsn can we get a lint for text outside of any section?

## Abstract

> Add a basic marketplace functionality to the [EIP-721](./eip-721.md)

Reviewer: ```suggestion
Add a basic marketplace functionality to [EIP-721](./eip-721.md)
```

> to to realize non-fungible tokens (NFTs) trading without relying on an NFTs trading intermediary platform.

Reviewer: ```suggestion
to enable non-fungible token trading without relying on an intermediary trading platform.
```

> It includes:
> - a method to list an item for sale or update an existing listing, whether private sale (only to a specific address) or public (to anyone),
> - a method to delist an item that has previously been listed,
> - a method to purchase a listed item,
> - a method to view all items listed for sale, and
> - a method to view a specific listing.

Reviewer: ```suggestion
```
This does not belong in the abstract

## Motivation

> Most of the current NFTs trading rely on the NFTs trading platform acting as an intermediary, which has the following problems.

Reviewer: ```suggestion
Most current NFT trading relies on an NFT trading platform acting as an intermediary, which has the following problems:
```

> 1.Security issues. For example, security issues caused by authorization through `setApprovalForAll`. The permission obtained by the NFT trading platform has carried out unnecessary risk exposure. Once there is a problem with the trading platform contract, it will bring huge losses to the entire industry.If the user has authorized the operation right of his/her NFTs to the NFT trading platform, the latest phishing website fraud method is to trick the user into signing, and let the victim place an order at a very low price on the NFTs trading platform, and designate the recipient. The scammer then accepts the deal, thereby reaping illicit benefits. Ordinary users are hard-pressed to guard against this.
> 
> 2.High trading costs. On the one hand, with the increase in the number of trading platforms, the liquidity of NFTs is scattered. If a user needs to make a deal as quickly as possible, he needs to authorize and place orders on multiple platforms, which further increases the risk exposure, and each authorization needs to spend gas. Taking BAYC as an example, the total number of BAYC is 10,000, and the number of current holders exceeds 6,000. The average number of BAYC held by each holder is less than 2. Although `setApprovalForAll` saves the subsequent pending order gas expenditure of a single trading platform, due to the need to authorize to Multi-platform, in essence, leads to an increase in gas expenditures for users. On the other hand, more importantly, the trading service fee charged by the trading platform must also be listed as the cost of user trading, which is much higher than the required gas spending for authorization.
> 
> 3.The aggregator provides a solution to aggregate liquidity, but whether the aggregator aggregates the liquidity of a certain trading platform, its decision is centralized. And because the order information of the trading platform is off-chain, the efficiency of the aggregator to obtain order data is affected by the frequency of the trading platform API, and sometimes the NFT trading platform will suspend the distribution of APIs and limit the frequency of APIs.
> 
> 4.Whether the NFT project party’s copyright tax income is obtained depends on the centralized decision-making of the NFT trading platform. Some trading platforms ignore the interests of the project party and implement 0 copyright tax, which infringes the interests of the project party. 
> 
> 5.Not resistant to censorship. Some platforms have delisted many NFTs, and the formulation and implementation of delisting rules are centralized and not transparent enough. Previously, some NFT trading platforms have also failed and wrongly delisted some NFTs, which caused market panic.

Reviewer: ```suggestion
1. Security issues. For example, security issues caused by authorization through `setApprovalForAll`. The permission obtained by the NFT trading platform has carried out unnecessary risk exposure. Once there is a problem with the trading platform contract, it will bring huge losses to the entire industry.If the user has authorized the operation right of his/her NFTs to the NFT trading platform, the latest phishing website fraud method is to trick the user into signing, and let the victim place an order at a very low price on the NFTs trading platform, and designate the recipient. The scammer then accepts the deal, thereby reaping illicit benefits. Ordinary users are hard-pressed to guard against this.
2. High trading costs. On the one hand, with the increase in the number of trading platforms, the liquidity of NFTs is scattered. If a user needs to make a deal as quickly as possible, he needs to authorize and place orders on multiple platforms, which further increases the risk exposure, and each authorization needs to spend gas. Taking BAYC as an example, the total number of BAYC is 10,000, and the number of current holders exceeds 6,000. The average number of BAYC held by each holder is less than 2. Although `setApprovalForAll` saves the subsequent pending order gas expenditure of a single trading platform, due to the need to authorize to Multi-platform, in essence, leads to an increase in gas expenditures for users. On the other hand, more importantly, the trading service fee charged by the trading platform must also be listed as the cost of user trading, which is much higher than the required gas spending for authorization.
3. The aggregator provides a solution to aggregate liquidity, but whether the aggregator aggregates the liquidity of a certain trading platform, its decision is centralized. And because the order information of the trading platform is off-chain, the efficiency of the aggregator to obtain order data is affected by the frequency of the trading platform API, and sometimes the NFT trading platform will suspend the distribution of APIs and limit the frequency of APIs.
4. Whether the NFT project party’s copyright tax income is obtained depends on the centralized decision-making of the NFT trading platform. Some trading platforms ignore the interests of the project party and implement 0 copyright tax, which infringes the interests of the project party. 
5. Not resistant to censorship. Some platforms have delisted many NFTs, and the formulation and implementation of delisting rules are centralized and not transparent enough. Previously, some NFT trading platforms have also failed and wrongly delisted some NFTs, which caused market panic.
```

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL
NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY",
and "OPTIONAL" in this document are to be interpreted as described in RFC 2119
and RFC 8174.

> Implementers of this standard MUST have all of the following functions:

Reviewer: ```suggestion
Compliant contracts MUST implement the following interface:
```

  /// @dev Returns a list of all current listings.
  ///
  /// @return the list of all currently listed tokens, 
  ///         along with their price and intended recipient
  ///
  function getAllListings() external view returns ( Listing[] memory );
  ///
  /// @dev Returns the listing for `tokenId`
  ///
  /// @return the specified listing (tokenId, price, intended recipient)
  ///
  function getListing( uint256 tokenId ) external view returns ( Listing memory );
}
```

The `listItem( uint256 tokenId, uint256 price, address to )` function MAY be implemented as `public` or `external`.

The `delistItem( uint256 tokenId )` function MAY be implemented as `public` or `external`.

The `buyItem( uint256 tokenId )` function MUST be implemented as `payable` and MAY be implemented as `public` or `external`.

The `getListing( uint256 tokenId )` function MAY be implemented as `pure` or `view`.

The `Listed` event MUST be emitted when an NFT is listed.

The `Delisted` event MUST be emitted when an NFT is delisted.

The `Purchased` event MUST be emitted when an NFT is traded.

The `supportsInterface` method MUST return `true` when called with `...`.

## Rationale

## Backwards Compatibility

> This standard is compatible with current [EIP-721](./eip-721.md) and [EIP-2981](./eip-2981.md) standards.

Reviewer: ```suggestion
This standard is compatible with [EIP-721](./eip-721.md) and [EIP-2981](./eip-2981.md).
```

    _tranfer( _tokenOwner_, msg.sender, tokenId_ );
    emit Purchased( tokenId_, _tokenOwner_, msg.sender, _listing_.price );

    // Handle royalties
    ( address _royaltyRecipient_, uint256 _royalties_ ) = royaltyInfo( tokenId_, msg.value );
    _processEthPayment( _royalties_, _royaltyRecipient_ );

    uint256 _payment_ = msg.value - _royalties_;
    _processEthPayment( _payment_, _tokenOwner_ );
  }

  // VIEW
  /**
  * @notice Returns the list of all existing listings.
  * 
  * @return the list of all existing listings
  */
  function getAllListings() external view returns ( Listing[] memory ) {
    return _listings;
  }

  /**
  * @notice returns the existing listing for `tokenId_`.
  * 
  * @return the existing listing for the requested token.
  */
  function getListing( uint256 tokenId ) external view returns ( Listing memory ) {
    return _listings[ _listingIndex[ tokenId_ ] ];
  }
}

   /// @dev See {IERC165-supportsInterface}.
   function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
     return interfaceId == type(IERC6105).interfaceId || super.supportsInterface(interfaceId);
        
```

## Security Considerations

> There are no security considerations related directly to the implementation of this standard.

Reviewer: ```suggestion
Needs discussion.
```
